lib/reservation_tracker.rb

```diff
@@ -0,0 +1,267 @@
+        block_qty_rooms = @rooms.take(requested_qty)
+        requested_dates = DateRange.new(start_date, end_date)
+
+        parsed_block = {
+          id: raw_block[:id].to_i,
+          party: block_qty_rooms,
+          date_range: requested_dates
+        }
+
+        block = Block.new(parsed_block)
+        blocks << block
+      end
+      return blocks
+    end
+
+    def load_reservations(filename)
+      reservations = []
+      reservation_data = CSV.open(filename, 'r', headers: true, header_converters: :symbol)
+
+      reservation_data.each do |raw_reservation|
+        block_id = find_block_id(raw_reservation[:block_id].to_i)
+        room = find_room(raw_reservation[:room].to_i)
+
+        start_date = Date.parse(raw_reservation[:start_date])
+        end_date = Date.parse(raw_reservation[:end_date])
+
+        range = DateRange.new(start_date, end_date)
+
+        parsed_reservation = {
+          id: raw_reservation[:id].to_i,
+          block_id: block_id,
+          room: room,
+          date_range: range
+        }
+
+        reservation = Reservation.new(parsed_reservation)
+        reservations << reservation
+      end
+      return reservations
+    end
+
+    def find_block_id(id)
+      check_id(id)
+      return @blocks.find { |block| block.id == id }
+    end
+
+    def find_room(room_num)
+      check_room_num(room_num)
+      return @rooms.find { |room| room.room_num == room_num }
+    end
+
+    def list_reservations_by_date(specified_date)
+      reservations_by_date = @reservations.find_all do |reservation|
+        date_range = reservation.date_range.get_range
+        date_range.include?(specified_date)
+      end
+      return reservations_by_date
+    end
+
+    def reservations_overlaps?(requested_dates)
+      matching_reservations = @reservations.find_all do |reservation|
+        date_range = reservation.date_range
+        reservation if date_range.overlaps?(requested_dates)
+      end
+    end
+
+    def find_reserved_rooms(requested_dates)
+      matching_reservations = reservations_overlaps?(requested_dates)
+
+      reserved_rooms = matching_reservations.map do |reservation|
+        reservation.room
+      end
+      return reserved_rooms
+    end
+
+    def find_blocks(requested_dates)
+      blocks_by_date = @blocks.map do |room|
+        date_range = room.date_range
+        room if date_range.overlaps?(requested_dates)
+      end
+      return blocks_by_date.compact
+    end
+
+    def find_unavailable_rooms(requested_dates)
+      reserved_rooms = find_reserved_rooms(requested_dates)
+      blocks = find_blocks(requested_dates)
+      unavailable_rooms = reserved_rooms + blocks
+      return unavailable_rooms
+    end
+
+    def find_available_rooms(requested_dates)
+      unavailable_rooms = find_unavailable_rooms(requested_dates)
+      available_rooms = @rooms.reject { |room| unavailable_rooms.include?(room) }
+      check_availablity?(available_rooms)
```

Comment:
You've done a great job throughout this file of breaking your code up into multiple small methods, each of which solves a small piece of the puzzle. Good work!